### ISSUE TITLE
Code refactor (11/16/15-present)

### ISSUE THREAD

**Issue by stephenfournier** [stephenfournier/dotastrikers#152]
## Code refactor (11/16/15-present)

[See my branch for the WIP](https://github.com/stephenfournier/dotastrikers/tree/myll)

The goals of the code refactor include:
* Utilizing the latest Barebones structure.
* Updating the code to use the latest custom game development idioms (i.e. removing all Scaleform, since that's going away very soon, probably after the Dota Majors.)
* Utilizing a lot more object orientation (lots of classes, don't repeat yourself, inheritence, etc).
* Transform all the physics to use the new physics library (not included in here yet.) This is more of a secondary goal since it's a big task in itself.

In vscripts/heroes you can see a good example of the OO I'll be implementing. There's a Hero class (hero.lua) which is like the base level class for a Dotastrikers hero, and then sub-classes like Antimage, Bloodseeker, etc. So, a hero unit can use the base functions in the Hero class, and if the base functions don't suffice, they can be overriden in the hero sub-class, like this: `if hero.Class.OnPhysicsFrame then return hero.Class:OnPhysicsFrame(hero) end`. Of course, the hero subclass can also provide unique functions to the hero. I also put the abilities unique to each hero in the corresponding hero folders, since I don't see any harm in doing that. This approach has worked pretty amazingly in my Dodgeball game.

This is obviously a really big refactor and it will probably take than a month or so to complete. @CarpeSwag I'm fine with doing it myself, you can continue adding features to the game or you can help me with the refactor, either way works for me.

**Comment by CarpeSwag** [comment]
I'm down to help, though I'd rather keep to the current folder layout rather than having a folder per hero.

I think I removed a lot of Scaleform already, but I'm not 100% sure where else there's scaleform.

**Comment by stephenfournier** [comment]
Yeah, that will work. We'll keep the abilities folder containing all the abilities, and remove having a folder per hero.

**Comment by stephenfournier** [comment]
There is scaleform where there is [this](https://github.com/stephenfournier/dotastrikers/search?utf8=%E2%9C%93&q=FireGameEvent)

**Comment by CarpeSwag** [comment]
Ah kk I'll go through the code and clean those out if you haven't already

**Comment by stephenfournier** [comment]
already did. haven't pushed it yet

**Comment by stephenfournier** [comment]
> For a lot of repeated code like that is there chance to use inheritance and define stuff once in the base class and have it infer the info from the subclasses?

Correct. `hero.Class = self` is necessary to achieve this. [Look at the current hero.lua](http://hastebin.com/uhenebopoy.lua), which is the baseclass for a hero (I have not pushed this yet). Look at the function `function Hero:OnBallControlled( hero )`. When I detect a new ball controller, this function is called. All heroes that become new ball controllers have this function called, regardless of specific subclass behavior. To achieve specific subclass behavior, notice this at the bottom: `if hero.Class.OnBallControlled then return hero.Class:OnBallControlled(hero) end`. So, if `Bloodseeker:OnBallControlled(hero)` were defined in the Bloodseeker class, then you would get Bloodseeker specific behavior when a bloodseeker hero becomes the new ball controller.

`if hero.Class.OnBallControlled then return hero.Class:OnBallControlled(hero) end` could have easily been placed at the top of `Hero:OnBallControlled( hero )` to achieve a completely overridden subclass function.

Stuff like `hero.isBloodseeker = true` is irrelevant to this, but still very useful to define (i.e. if you have some function in gamemode.lua that needs to do hero-specific behavior). I could have done `if hero:GetClassname() == "npc_dota_hero_bloodseeker then` instead of `if hero.isBloodseeker then` to achieve the same effect, but that requires more typing in the long run.
